src/static/helpers/slasHelper.ts

```diff
@@ -0,0 +1,258 @@
+ * @returns code verifier
+ */
+export const createCodeVerifier = (): string => nanoid(128);
+
+/**
+ * Encodes a code verifier to a code challenge to send to the authorization endpoint
+ * @param codeVerifier random string to use as a code verifier
+ * @returns code challenge
+ */
+export const generateCodeChallenge = async (
+  codeVerifier: string
+): Promise<string> => {
+  const urlSafe = (input: string) =>
+    input.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
+
+  let challenge = '';
+  try {
+    if (window.crypto.subtle.digest !== undefined) {
+      const encoder = new TextEncoder();
+      const data = encoder.encode(codeVerifier);
+      const digest = await window.crypto.subtle.digest('SHA-256', data);
+      const base64Digest = btoa(String.fromCharCode(...new Uint8Array(digest)));
+      challenge = urlSafe(base64Digest);
+    }
+  } catch (e) {
+    const crypto = await import('crypto');
+    challenge = urlSafe(
+      crypto.default.createHash('sha256').update(codeVerifier).digest('base64')
+    );
+  }
+
+  if (challenge.length === 0) {
+    throw new Error('Problem generating code challenge');
+  }
+
+  return challenge;
+};
+
+/**
+ * Wrapper for the authorization endpoint. For federated login (3rd party IDP non-guest), the caller should redirect the user to the url in the url field of the returned object. The url will be the login page for the 3rd party IDP and the user will be sent to the redirectURI on success. Guest sessions return the code and usid directly with no need to redirect.
+ * @param slasClient a configured instance of the ShopperLogin SDK client
+ * @param parameters -
+ *   codeVerifier - random string created by client app to use as a secret in the request
+ *   redirectURI - the location the client will be returned to after successful login with 3rd party IDP. Must be registered in SLAS.
+ *   hint? - optional string to hint at a partitcular IDP. Guest sessions are created by setting this to 'guest'
+ *   usid? - optional saved SLAS user id to link the new session to a previous session
+ * @returns login url, user id and authorization code if available
+ */
+export async function authorize(
+  slasClient: ShopperLogin<{
+    shortCode: string;
+    organizationId: string;
+    clientId: string;
+  }>,
+  parameters: {
+    codeVerifier: string;
+    redirectURI: string;
+    hint?: string;
+    usid?: string;
+  }
+): Promise<Record<string, string>> {
+  const codeChallenge = await generateCodeChallenge(parameters.codeVerifier);
+
+  const options = {
+    parameters: {
+      client_id: slasClient.clientConfig.parameters.clientId,
+      code_challenge: codeChallenge,
+      ...(parameters.hint && {hint: parameters.hint}),
+      organizationId: slasClient.clientConfig.parameters.organizationId,
+      redirect_uri: parameters.redirectURI,
+      response_type: 'code',
+      ...(parameters.usid && {usid: parameters.usid}),
```

Comment:
Some of this may point to things that need tweaking in the generation at some point, but I get not assignable for `authorizeCustomer` without explicitly setting `clientId` and `organizationId`. Passing `usid` and `hint` were causing problems at runtime because the query params were being sent empty and the service was complaining (I know that shouldn't be the case looking at the code, but that's what was happening.)